src/nvim/terminal.c

```diff
@@ -496,37 +495,33 @@ static int terminal_execute(VimState *state, int key)
 void terminal_destroy(Terminal *term)
 {
   buf_T *buf = handle_get_buffer(term->buf_handle);
-  if (buf) {
-    term->buf_handle = 0;
-    buf->terminal = NULL;
-  }
+  assert(buf);
+  apply_autocmds(EVENT_TERMCLOSE, NULL, NULL, false, buf);
```

Comment:
Then nvim would block until the process gets killed.
I think having `TermClose` trigger after the process exits would be best, considering that there are other events in place that trigger when the buffer gets closed (`BufDelete`, `BufWipeout`).
Then again, changing the `TermClose` behavior could break compatibility with existing plugins (the current behavior is the same as for `HEAD`).